internal/streamingnode/server/service/handler/producer/produce_server.go

```diff
@@ -0,0 +1,234 @@
+			p.logger.Warn("unknown request type", zap.Any("request", req))
+		}
+	}
+}
+
+// handleProduce handles the produce message request.
+func (p *ProduceServer) handleProduce(req *streamingpb.ProduceMessageRequest) {
+	p.logger.Debug("recv produce message from client", zap.Int64("requestID", req.RequestId))
+	msg := message.NewMutableMessageBuilder().
+		WithPayload(req.GetMessage().GetPayload()).
+		WithProperties(req.GetMessage().GetProperties()).
+		BuildMutable()
+
+	if err := p.validateMessage(msg); err != nil {
+		p.logger.Warn("produce message validation failed", zap.Int64("requestID", req.RequestId), zap.Error(err))
+		p.sendProduceResult(req.RequestId, nil, err)
+		return
+	}
+
+	// Append message to wal.
+	// Concurrent append request can be executed concurrently.
+	messageSize := msg.EstimateSize()
+	now := time.Now()
+	p.appendWG.Add(1)
+	p.wal.AppendAsync(p.produceServer.Context(), msg, func(id message.MessageID, err error) {
+		defer func() {
+			p.appendWG.Done()
+			p.updateMetrics(messageSize, time.Since(now).Seconds(), err)
+		}()
+		p.sendProduceResult(req.RequestId, id, err)
+	})
+}
+
+// validateMessage validates the message.
+func (p *ProduceServer) validateMessage(msg message.MutableMessage) error {
+	// validate the msg.
+	if !msg.Version().GT(message.VersionOld) {
+		return status.NewInner("unsupported message version")
+	}
+	if !msg.MessageType().Valid() {
+		return status.NewInner("unsupported message type")
+	}
+	if msg.Payload() == nil {
+		return status.NewInner("empty payload for message")
+	}
+	return nil
+}
+
+// sendProduceResult sends the produce result to client.
+func (p *ProduceServer) sendProduceResult(reqID int64, id message.MessageID, err error) {
+	resp := &streamingpb.ProduceMessageResponse{
+		RequestId: reqID,
+	}
+	if err != nil {
+		p.logger.Warn("append message to wal failed", zap.Int64("requestID", reqID), zap.Error(err))
+		resp.Response = &streamingpb.ProduceMessageResponse_Error{
+			Error: status.AsStreamingError(err).AsPBError(),
+		}
+	} else {
+		resp.Response = &streamingpb.ProduceMessageResponse_Result{
+			Result: &streamingpb.ProduceMessageResponseResult{
+				Id: &streamingpb.MessageID{
+					Id: id.Marshal(),
+				},
+			},
+		}
+	}
+
+	// If server context is canceled, it means the stream has been closed.
+	// all pending response message should be dropped, client side will handle it.
+	select {
+	case p.produceMessageCh <- resp:
+		p.logger.Debug("send produce message response to client", zap.Int64("requestID", reqID), zap.Any("messageID", id), zap.Error(err))
+	case <-p.produceServer.Context().Done():
+		p.logger.Warn("stream closed before produce message response sent", zap.Int64("requestID", reqID), zap.Any("messageID", id))
+		return
+	}
+}
+
+// updateMetrics updates the metrics.
+func (p *ProduceServer) updateMetrics(messageSize int, cost float64, err error) {
+	name := p.wal.Channel().Name
+	term := strconv.FormatInt(p.wal.Channel().Term, 10)
+	if err == nil {
```

Comment:
Here should include statistics of size and latency for produce requests if no errors.